### PR TITLE
Allow running mastodon on a different domain as the one used for identifying users

### LINKED ISSUE #1032: Subdomains
I haven't gotten to running my own instance yet (soon!), but I'm seeing a lot of `@username@mastodon.example.com` in toots, instead of `@username@example.com`.

With email, you can host a server and access your mail at `mail.example.com` but still address to `username@example.com` thanks to MX DNS records. Users get a simple email address, and the servers figure out the rest.

If you have to use a subdomain because you already have something else on at `example.com`, it seems like you end up having to use  `username@mastodon.example.com` as your username instead, unless you create a new domain, like `example.social`.

It looks like there's a `LOCAL_DOMAIN` variable, but I haven't seen anything akin to email's MX record. I can't find much documentation on `LOCAL_DOMAIN`. I'm not sure if this is already there, but hasn't been fully documented/communicated yet, or this just hasn't happened yet.

* * * *

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

### PR DESCRIPTION
**Pull request by ClearlyClaire**
This should fix #1032, hopefuly.

### CODE REVIEW

**Comment by ThibG** [comment]
With this change, having the following in .env.production allows to have user@example.org handled by mastodon running on social.example.org, as long as https://example.org/.well-known/host-meta redirects to or proxies https://social.example.org/.well-known/host-meta

> LOCAL_DOMAIN=example.org
> WEB_DOMAIN=social.example.org

File: .env.production.sample [hunk 1]
```diff
@@ -11,6 +11,9 @@ DB_PORT=5432
 LOCAL_DOMAIN=example.com
 LOCAL_HTTPS=true
 
+# Use this if mastodon has to run on a different domain as the one used for federation
+#WEB_DOMAIN=mastodon.example.com
```
Comment (Gargron) [inline .env.production.sample]: I would add a note that people should leave it commented out unless they know exactly what they are doing - and add a space after `#` for consistency.

**Review by Gargron — CHANGES_REQUESTED** [review]
:+1:

**Comment by ThibG** [comment]
Actually, something might still be amiss with PubSubHubbub subscriptions… while two people followed me since I've re-set my instance, there doesn't seem to be any effective subscription, and they don't see my toots. People from that server where already following me before I re-set my instance, and the feeds URL has stayed the same, which may or may not explain the issue.
I'll investigate further…

**Comment by ThibG** [comment]
It seems that even on instances that don't have faulty URLs in their database (#1276), PubSubHubbub fail with HTTP status 422. I don't know why yet.

**Comment by Gargron** [comment]
Ah, I think I know why that might be. PubSubHubbub validates that the requested feed URL belongs to your domain. I need to double check which variable it uses to check the domain.

**Comment by Gargron** [comment]
Indeed, `Api::PushController#topic_to_account` checks the domain with `TagManager.instance.local_domain?` which is otherwise used only for federation purposes. Perhaps worth it to just do the check locally in that controller instead, it just needs to be a case-insensitive compare against WEB_DOMAIN || LOCAL_DOMAIN

**Comment by ThibG** [comment]
Thanks for the hint, my last patch fixed it :)
Now, I only have to deal with #1276 !

File: app/services/process_feed_service.rb [hunk 2]
```diff
@@ -163,7 +163,7 @@ def mentions_from_xml(parent, xml)
 
         url = Addressable::URI.parse(link['href'])
 
-        mentioned_account = if TagManager.instance.local_domain?(url.host)
+        mentioned_account = if TagManager.instance.web_domain?(url.host)
```
Comment (Gargron) [inline app/services/process_feed_service.rb]: No no no, this is the opposite of what you want. Web domain = mastodon.example.com. You want your users to be on example.com. Revert this one

**Review by Gargron — CHANGES_REQUESTED** [review]

File: app/services/process_feed_service.rb (hunk 2 above)
Comment (ThibG) [inline app/services/process_feed_service.rb]: Indeed, misread this part, sorry about that! I have reverted it.

Comment (ThibG) [inline app/services/process_feed_service.rb]: Actually no, the URI in the “mentioned” links from other people's toots seem to be full URLs to the web front-end, thus using WEB_DOMAIN and not LOCAL_DOMAIN

**Review by Gargron — APPROVED** [review]

**Comment by plinss** [comment]
This change is much appreciated. Question though: is it safe to update LOCAL_DOMAIN/WEB_DOMAIN on an already running instance? e.g. if I change from:
```
LOCAL_DOMAIN=social.example.com
```
to:
```
LOCAL_DOMAIN=example.com
WEB_DOMAIN=social.example.com
```
(and have the webfinger proxied from example.com, of course)
Will existing users on other instances who are already following @user@social.example.com still be following properly? Or do they all have to start over with @user@example.com?

**Comment by ThibG** [comment]
I am sorry, but this change is not completely usable yet: indeed, remote Mastodon instances will fail at accepting salmon API requests (follow/unfollow notifications, private messages, any message if you're not followed by someone on the remote instance) from you (see #2012).

As for your migration question, this is tricky. The upside is that you will keep all your toots, and referencing them should still work. You will also keep your profile URL (at social.example.com) and URI.

However, your remote contacts will still be following user@social.example.com and think they are followed by user@social.example.com. So far, it's just standard migration issues. But worse, interaction with user@social.example.com will be broken (so you can't easily tell people to switch over to user@example.com, you won't have both at the same time). Even worse, remote instances that knew social.example.com will already have your URI associated to the obsolete user@social.example.com, which might break salmon requests and mentions processing even when #2012 gets fixed.
With this two issues, “starting over” won't help you, as the important thing here is the URI, which will be the same whether you're “starting other” or not. This is related to #2032 except in this case, the URI was “valid” at the moment the user was registered in both cases, I'm not sure how it should be handled…

**Comment by michaeljdeeb** [comment]
This might be a niche use case, but I wanted to point it out. I'm currently hosting a static site at my root domain using GitHub Pages and a Mastodon instance at [mastodon.michaeljdeeb.com](https://mastodon.michaeljdeeb.com). I'm only able to use an http refresh to do the redirect and because there's no extension on host-meta the meta redirect file is downloaded instead of actually redirecting.

The workaround is to just host the root site myself, but if this is a solvable problem you might be helping out many other single user instances in the same situation.

**Comment by plinss** [comment]
@michaeljdeeb if you have control of your DNS, then a SRV record (and support in Mastodon/the GnuSocial protocols) would solve your use case.

**Comment by ThibG** [comment]
@michaeljdeeb I guess you can simply host the host-meta file on your static website without doing any redirection. This host-meta file will mention mastodon.michalejdeeb.com for webfinger queries, and that's it!

But as I've said earlier, remote Mastodon instances will currently fail to accept salmon requests from your instance, thus preventing you from interacting with private accounts and instances where noone follows you (I think).

**Comment by michaeljdeeb** [comment]
Oh! I did try that, but it gave me errors when I would try to follow people or they me, but now I'm thinking that might have to do with an HTTP root domain and HTTPS Mastodon instance.

**Pull request merged**